### PR TITLE
Improved VMRay CA

### PR DESCRIPTION
**Pull request by 53A-1**
This PR contains a major refactor of the VMRay Cortex Analyzer and a lot of new features. C.f. the commit message for a list of changes.
A few remarks on those which are not self-explanatory:
- Implemented recursion of samples (relations) and limits on it
    - VMRay supports recursively analyzing observables which have been discovered during the analysis of the main submission. This might not be the desired behavior for TheHive, since the threat scoring might not be comparable to the results produced by other Cortex Analyzers. However, this functionality provides a more thorough view on the submission and is therefore enabled by default
- Adding support for archives with passwords
and
- Adding support for different archive modes (compound and separate)
    - This has been added in case users want to directly submit archives using Cortex without TheHive extracting them
- Formatting code using black (https://github.com/psf/black)
    - I did not find a default formatter for the project, hope this one is fine with you

This implementation is based on [Pull request #802](https://github.com/TheHive-Project/Cortex-Analyzers/pull/802) by [mback2k](https://github.com/mback2k)

### CODE REVIEW

**Comment by mback2k** [comment]
@jeromeleonard why was this just closed away with a merge into any development branch?

**Comment by mback2k** [comment]
@53A-1 thanks for your work, I will test it in our infrastructure this week. I may have some additional enhancements to follow up.

One thing for a potential recreation of this PR: I would recommend keeping my commits in the history and also splitting your work into logical sections, because at the moment it is a bit hard to review to the huge amount of changes in a single commit.

File: analyzers/VMRay/vmray.py [hunk 1]
```diff
@@ -7,98 +7,230 @@
...
+        "malicious": "malicious",
+        "blacklisted": "malicious",
+    }
+
+    _ioc_mapping = {
+        "domains": "domain",
+        "email_addresses": "email",
+        "emails": "sender",
+        "files": "filename",
+        "ips": "ip_address",
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:27]: Cortex uses just `ip` for this data type instead of `ip_address`.

File: analyzers/VMRay/vmray.py [hunk 2]
```diff
@@ -7,98 +7,230 @@
...
+    _severity_mapping = {
+        "whitelisted": "safe",
+        "suspicious": "suspicious",
+        "malicious": "malicious",
+        "blacklisted": "malicious",
+    }
+
+    _ioc_mapping = {
+        "domains": "domain",
+        "email_addresses": "email",
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:24]: Cortex uses just `mail` for this data type instead of `email`.

File: analyzers/VMRay/vmray.py [hunk 3]
```diff
@@ -7,98 +7,230 @@
...
+        "emails": "sender",
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:25]: What is the difference to `emails`? I think this should also just map to `mail`.

File: analyzers/VMRay/vmray.py [hunk 4]
```diff
@@ -7,98 +7,230 @@
...
+        "files": "filename",
+        "ips": "ip_address",
+        "mutexes": "mutex_name",
+        "registry": "reg_key_name",
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:29]: Cortex uses just `registry` for this data type instead of `reg_key_name`.

File: analyzers/VMRay/vmray.py [hunk 5]
```diff
@@ -7,98 +7,230 @@
...
+        "blacklisted": "malicious",
+    }
+
+    _ioc_mapping = {
+        "domains": "domain",
+        "email_addresses": "email",
+        "emails": "sender",
+        "files": "filename",
+        "ips": "ip_address",
+        "mutexes": "mutex_name",
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:28]: Cortex does not know about `mutex_name`. So I guess `other` would be the right data type.

File: analyzers/VMRay/vmray.py [hunk 6]
```diff
@@ -7,98 +7,230 @@
...
+                if ioc_type in iocs:
+                    for ioc_node in iocs[ioc_type]:
+                        severity = ioc_node.get("severity", "")
+                        level = self._severity_mapping.get(severity, "info")
+                        tags = [severity, level, ioc_node["type"]]
+                        payload = ioc_node[ioc_payload_name]
+                        if "hashes" in ioc_node:
+                            for hash_node in ioc_node["hashes"]:
+                                if "sha256_hash" in hash_node:
+                                    payload += ", SHA256: {}".format(
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:212]: The `payload` must not contain any meta data. It must be a valid artifact for re-analysis, eg. just an URL or just a hash. This currently breaks recursive analyzing logic! Metadata can be put into the `message` or `tags` of the artifact.

File: analyzers/VMRay/vmray.py [hunk 7]
```diff
@@ -7,98 +7,230 @@
...
+                                        hash_node["sha256_hash"]
+                                    )
+                        elif "operations" in ioc_node:
+                            payload += ", Operations: " + ", ".join(
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:216]: Same as above. Use `message` or `tags` for this.

File: analyzers/VMRay/vmray.py [hunk 8]
```diff
@@ -7,98 +7,230 @@
...
+                                ioc_node["operations"]
+                            )
+                        artifacts.append(
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:219]: A single hash or a single URL maps to one artifact, eg. an artifact itself does not contain a list as payload. Artifacts is a list of hashes, URLs, etc. with some meta data attached to them.

File: analyzers/VMRay/vmray.py [hunk 9]
```diff
@@ -7,98 +7,230 @@
...
+        for sample in samples:
+            link = sample.get("sample_webif_url", None)
+            iocs = sample.get("sample_iocs", {}).get("iocs", {})
+
+            for ioc_type, ioc_payload_name in self._ioc_mapping.items():
+                if ioc_type in iocs:
+                    for ioc_node in iocs[ioc_type]:
+                        severity = ioc_node.get("severity", "")
+                        level = self._severity_mapping.get(severity, "info")
+                        tags = [severity, level, ioc_node["type"]]
```
Comment (mback2k) [inline analyzers/VMRay/vmray.py:207]: My original PR used a `set` on purpose to avoid duplicate tags since `severity` can times be equal to `level`.

**Comment by dadokkio** [comment]
Sorry @53A-1 @mback2k, 
we didn't see these pulls on 2.8.0 branch and when we released the milestone they have been automatically closed.

It'll be possible for you to make a new pull based on develop?

**Comment by mback2k** [comment]
Yes, I am already preparing some changes based upon my review for @53A-1 to review and include in his next PR then here:
https://github.com/TKCERT/Cortex-Analyzers/commits/develop-vmray

**Pull request closed without merging**